inmoose/limma/lmfit.py

```diff
@@ -0,0 +1,433 @@
+        to samples and columns to coefficients to be estimated.
+        Defaults to :code:`obj.design` if not :code:`None`, otherwise to the
+        unit vector, meaning that all samples will be treated as replicated of
+        a single treatment group.
+    ndups : int
+        positive integer giving the number of times each distinct probe is
+        printed on each array
+    spacing : int
+        positive integer giving the spacing between duplicate occurrences of
+        the same probe, :code:`spacing=1` for consecutive rows.
+    block : array-like
+        vector or factor specifying a blocking variable on the arrays. Has
+        length equal to the number of arrays. Must be :code:`None` if
+        :code:`ndups > 2`.
+    correlation
+        the inter-duplicate or inter-technical replicate correlation
+    weights
+        non-negative precision weights. Can be a numeric matrix of individual
+        weights of same size as the object expression matrix, or a numeric
+        vector of gene weights with length equal to :code:`nrow` of the
+        expression matrix.
+    method : { "ls", "robust" }
+        fitting method: :code:`"ls"` for least squares or :code:`"robust"` for
+        robust regression
+
+    Returns
+    -------
+    MArrayLM
+        object containing the result of the fits.
+        The row names of :code:`obj` are preserved in the fit object and can be
+        retrieved by :code:`fit.index` where :code:`fit` is the output of
+        :fun:`lmFit`.
+        The column names of :code:`design` are preserved as column names and
+        can be retrieved by :code:`fit.columns`.
+    """
+    # Fit genewise linear models
+
+    # Extract components from obj
+    y = EAWP(obj)
+
+    if y.exprs.shape[0] == 0:
+        raise ValueError("expression matrix has zero rows")
+
+    # Check design matrix
+    if design is None:
+        design = y.design
+    if design is None:
+        design = patsy.dmatrix("~1", data=y.exprs)
+    else:
+        if not isinstance(design, patsy.DesignMatrix):
+            raise ValueError("design must be a patsy DesignMatrix")
+        if design.shape[0] != y.exprs.shape[1]:
+            raise ValueError(
+                "row dimension of design does not match column dimension of data object"
+            )
+        if np.isnan(design).any():
+            raise ValueError("NAs not allowed in design matrix")
+
+    ne = nonEstimable(design)
+    if ne is not None:
+        logging.info(f"Coefficients not estimable: {' '.join(ne)}")
+
+    # Check ndups and spacing. Default to 1.
+    if ndups is None:
+        try:
+            ndups = y.printer.ndups
+        except AttributeError:
+            pass
+    if ndups is None:
+        ndups = 1
+    if spacing is None:
+        try:
+            spacing = y.printer.spacing
+        except AttributeError:
+            pass
+    if spacing is None:
+        spacing = 1
```

Comment:
Why do we check twice if `spacing is None` with 2 different following action? Can this `spacing is None` only happen if`y.printer.spacing is None`, so perhaps we can include this code in the previous if?

Reply:
The scenario is as follows:

1. if `spacing` is `None` then we **try** to retrieve its value from `y.printer`. Several cases are to be considered:
   a. `y.printer` is absent -- in which case `spacing` remains `None`
   b. `y.printer` is present but it has no attribute `spacing` -- in which case `spacing` remains `None`
   c. `y.printer.spacing` is `None` -- in which case `spacing` is still `None`
   d. `y.printer.spacing` is not `None`
2. after that, if `spacing` is still `None` (cases a. b. c. above), then use the default value 1

In current implementation, the `except` clause accounts for cases a. and b., while the second `if` differentiates between (a. or b. or c.) and (d.).

Reply:
Very clear thanks !